src/translators/typescript.py

```diff
@@ -0,0 +1,748 @@
+
+class TypeScriptTranslator(BaseTranslator):
+    filename = "Main.ts"
+    incorrect_filename = "Incorrect.ts"
+    executable = "Main.js"
+    ident_value = " "
+
+    def __init__(self, package=None, options={}):
+        super().__init__(package, options)
+        self._children_res = []
+        self.ident = 0
+        self.is_interface = False
+        self.is_void = False
+        self.is_lambda = False
+        self.current_class = None
+        self.current_function = None
+        self.context = None
+        self._namespace: tuple = ast.GLOBAL_NAMESPACE
+        self._nodes_stack = [None]
+    
+    def _reset_state(self):
+        self._children_res = []
+        self.ident = 0
+        self.is_interface = False
+        self.is_void = False
+        self.is_lambda = False
+        self.current_class = None
+        self.current_function
+        self.context = None
+        self._namespace = ast.GLOBAL_NAMESPACE
+        self._nodes_stack = [None]
+    
+    @staticmethod
+    def get_filename():
+        return TypeScriptTranslator.filename
+    
+    @staticmethod
+    def get_incorrect_filename():
+        return TypeScriptTranslator.incorrect_filename
+    
+    def type_arg2str(self, t_arg):
+        # TypeScript does not have a Wildcard type
+        a = self.get_type_name(t_arg)
+        return self.get_type_name(t_arg)
+    
+    def get_type_name(self, t):
+        t_constructor = getattr(t, 't_constructor', None)
+        if not t_constructor:
+            return t.get_name()
+
+        if t_constructor.name.startswith("Function"):
+            param = ['p', 1]
+            res = "("
+
+            if len(t.type_args) == 1:
+                res += ") => " + self.type_arg2str(t.type_args[-1])
+                return res
+            
+            for ta in t.type_args[:-1]:
+                res += param[0] + str(param[1]) + ": " + self.type_arg2str(ta) + ", "
+                param[1] += 1
+            res = res[:-2] # removes , from end of string
+            res += ") => " + self.type_arg2str(t.type_args[-1])
+            return res
+
+        return "{}<{}>".format(t.name, ", ".join([self.type_arg2str(ta)
+                                                  for ta in t.type_args]))
+
+    def pop_children_res(self, children):
+        len_c = len(children)
+        if not len_c:
+            return []
+        res = self._children_res[-len_c:]
+        self._children_res = self._children_res[:-len_c]
+        return res
+    
+    def visit_program(self, node):
+        self.context = node.context
+        self.class_decls = [decl for decl in node.declarations
+                            if isinstance(decl, ast.ClassDeclaration)]
+        children = node.children()
+        for c in children:
+            c.accept(self)
+        res = '\n\n'.join(self.pop_children_res(children))
+        self.program = (
+            res if self.package is None
+            else "module {} {{\n{}\n}}".format(self.package, res)
+        )
```

Comment:
call `self._reset_state()` at this point. In this way, if we want to translate a new program using the same visitor, it is going to use a fresh state, and not the state from the previous translation.